<a href="https://colab.research.google.com/github/monda00/horse-race-notebook/blob/master/predict_show_neural_network.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ニューラルネットワークで予測

学習データの作成から予測の考察までやってみる。

- 単純に１行ずつ学習
- 各レースのデータ数を揃えて１レースまとめて学習

特徴量

|分類	|項目 |
|---|---|
|馬情報	|馬番 |
| |枠番 |
| |年齢 |
| |性別 |
| |体重（現在） | 
| |体重（前走との差分） |
| |負担重量 |
| 当日レース情報 |レース場 |
| |出走馬数 |
| |コース距離 |
| |周回方向 |
| |コースタイプ（ダ/芝/障） |
| |天気 |
| |馬場状態 |
| |開始時間帯 |
| |時期 |
|同馬の過去レース情報（×5走分）	|オッズ |
| |人気 |
| |順位 |
| |タイム（秒） |
| |前走からの経過日数 |
| |コース距離 |
| |コースタイプ（ダ/芝/障） |
| |天気 |
| |馬場状態 |

# 概要

- ライブラリ・データ読み込み
- 前処理
- 学習
- 予測
- 考察

## 参考

- [データ収集からディープラーニングまで全て行って競馬の予測をしてみた](https://qiita.com/kami634/items/55e49dad76396d808bf5#%E5%8F%96%E5%BE%97%E3%81%97%E3%81%9Furl%E3%82%92%E3%82%82%E3%81%A8%E3%81%ABhtml%E3%82%92%E5%BE%97%E3%82%8B)
- [競馬の予測をガチでやってみた](http://stockedge.hatenablog.com/entry/2016/01/03/103428)
- [ディープラーニングさえあれば、競馬で回収率100%を超えられる](https://qiita.com/yossymura/items/334a8f3ef85bff081913)
- [競馬予想AIを作る 〜ニューラルネットワークによる相対評価データセットの取り扱い例〜](https://cocon-corporation.com/cocontoco/horseraceprediction_ai/)

# ライブラリ・データ読み込み

In [1]:
import numpy as np
import pandas as pd
import datetime
from tqdm import tqdm
import collections

from tensorflow import keras
from tensorflow.keras.layers import Input, Dense, Dropout, BatchNormalization
from tensorflow.keras.layers import concatenate
from tensorflow.keras.models import Model

from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import train_test_split

In [2]:
DATA_PATH = '/content/drive/My Drive/data/horse-race/'

In [3]:
df = pd.read_csv(DATA_PATH + 'train_nn.csv')

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (5,6,32,33,34,41,42,43,50,51,52,59,60,61) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [4]:
df.head()

race_id  horse_number  frame_number  ...  five_before_weather show      date
0  201945010102           1.0           1.0  ...                    0  0.0  2019/1/1
1  201945010102           2.0           2.0  ...                    0  1.0  2019/1/1
2  201945010102           3.0           3.0  ...                    0  1.0  2019/1/1
3  201945010102           4.0           4.0  ...                    0  0.0  2019/1/1
4  201945010102           5.0           5.0  ...                    0  1.0  2019/1/1

[5 rows x 64 columns]

# 前処理

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 407340 entries, 0 to 407339
Data columns (total 64 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   race_id                       407340 non-null  int64  
 1   horse_number                  253636 non-null  float64
 2   frame_number                  253636 non-null  float64
 3   age                           253636 non-null  float64
 4   gen                           253636 non-null  object 
 5   weight                        253636 non-null  object 
 6   weight_diff                   253636 non-null  object 
 7   burden_weight                 253636 non-null  float64
 8   place                         253636 non-null  object 
 9   race_horse_number             253636 non-null  float64
 10  distance                      253636 non-null  float64
 11  clockwise                     253636 non-null  object 
 12  field_type                    253636 non-nul

## 日付削除

日付はソートのために利用するだけのため削除

In [6]:
df = df.drop('date', axis=1)

## 欠損値

0埋めする。

In [7]:
df.isnull().sum().sum()

9529648

In [8]:
df = df.fillna(0)

過去レースの情報がない馬も結構いるかもしれない。

23752個のデータで過去レースの情報がない。
数値と文字列の0が混在している。

In [9]:
past_race_columns_base = ['odd', 'popular', 'rank', 'time', 'elapsed_day', 'distance', 'field_type', 'field_condition', 'weather']
past_race_num = ['one', 'two', 'three', 'four', 'five']
past_race_columns = []
for n in past_race_num:
  for c in past_race_columns_base:
    past_race_columns.append('{}_before_{}'.format(n, c))

## weight

計測不能が混じっている。

In [10]:
len(df[df['weight'] == '計不'])

44

In [11]:
df[df['weight'] == '計不'] = 0

In [12]:
df['weight'] = df['weight'].astype('int64')

## weight diff

型を変換する

In [13]:
df['weight_diff'] = df['weight_diff'].astype('int64')

## Label Encoding

In [14]:
categorical_cols = ['gen', 'place', 'clockwise', 'field_type', 'field_condition', 'weather', 'season']
categorical_cols_past_base = ['field_type', 'field_condition', 'weather']

In [15]:
categorical_cols_past = []
for n in past_race_num:
  for c in categorical_cols_past_base:
    categorical_cols_past.append('{}_before_{}'.format(n, c))

In [16]:
categorical_cols = categorical_cols + categorical_cols_past

In [17]:
df[categorical_cols] = df[categorical_cols].astype(str)

In [18]:
for c in categorical_cols:
  le = LabelEncoder()
  le.fit(df[c])
  df[c] = le.transform(df[c])

In [19]:
df

race_id  horse_number  ...  five_before_weather  show
0       201945010102           1.0  ...                    1   0.0
1       201945010102           2.0  ...                    1   1.0
2       201945010102           3.0  ...                    1   1.0
3       201945010102           4.0  ...                    1   0.0
4       201945010102           5.0  ...                    1   1.0
...              ...           ...  ...                  ...   ...
407335  202048060911           0.0  ...                    0   0.0
407336  202048060911           0.0  ...                    0   0.0
407337  202048060911           0.0  ...                    0   0.0
407338  202048060911           0.0  ...                    0   0.0
407339  202048060911           0.0  ...                    0   0.0

[407340 rows x 63 columns]

# 学習

## 学習データと検証データに分割

In [20]:
train_df, test_df = train_test_split(df, test_size=54000, shuffle=False)

In [21]:
test_df

race_id  horse_number  ...  five_before_weather  show
353340  202050042307           1.0  ...                    4   0.0
353341  202050042307           2.0  ...                    4   1.0
353342  202050042307           3.0  ...                    4   0.0
353343  202050042307           4.0  ...                    4   1.0
353344  202050042307           5.0  ...                    4   1.0
...              ...           ...  ...                  ...   ...
407335  202048060911           0.0  ...                    0   0.0
407336  202048060911           0.0  ...                    0   0.0
407337  202048060911           0.0  ...                    0   0.0
407338  202048060911           0.0  ...                    0   0.0
407339  202048060911           0.0  ...                    0   0.0

[54000 rows x 63 columns]

In [22]:
X_train = train_df.drop(['race_id', 'show'], axis=1).values
y_train = train_df['show'].values
X_test = test_df.drop(['race_id', 'show'], axis=1).values
y_test = test_df['show'].values  

In [23]:
train_race_id_counter = collections.Counter(list(train_df['race_id'].values))
test_race_id_counter = collections.Counter(list(test_df['race_id'].values))
train_query = list(train_race_id_counter.values())
test_query = list(test_race_id_counter.values())

## 正規化

単純な標準化をする。

nullを0で補完しているためスケールがバラバラになってそう。

In [24]:
sc = StandardScaler()
sc.fit(X_train)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [25]:
X_train_sc = sc.transform(X_train)
X_test_sc = sc.transform(X_test)

In [26]:
print('train X', X_train_sc.shape)
print('train y', y_train.shape)
print('test X', X_test_sc.shape)
print('test y', y_test.shape)

train X (353340, 61)
train y (353340,)
test X (54000, 61)
test y (54000,)


In [27]:
X_train_sc = X_train_sc.reshape([-1, 18, 61])
y_train = y_train.reshape([-1, 18])
X_test_sc = X_test_sc.reshape([-1, 18, 61])
y_test = y_test.reshape([-1, 18])

In [28]:
print('train X', X_train_sc.shape)
print('train y', y_train.shape)
print('test X', X_test_sc.shape)
print('test y', y_test.shape)

train X (19630, 18, 61)
train y (19630, 18)
test X (3000, 18, 61)
test y (3000, 18)


### 多入力モデル用に整形

In [29]:
X_train_sc_reshape = []
X_test_sc_reshape = []
for i in range(18):
  X_train_sc_reshape.append(X_train_sc[:, i, :])
  X_test_sc_reshape.append(X_test_sc[:, i, :])

## モデルの定義

In [30]:
def define_simple_model(input_shape):
  inp = Input(shape=input_shape)
  x = Dense(300, activation='relu')(inp)
  x = Dropout(0.2)(x)
  x = Dense(150, activation='relu')(x)
  x = Dropout(0.2)(x)
  x = Dense(50, activation='relu')(x)
  x = Dropout(0.2)(x)
  x = Dense(1, activation='sigmoid')(x)

  model = Model(inp, x)
  model.compile(optimizer='adam', loss='mean_squared_error')

  return model

In [31]:
def define_race_model(input_shape):
  horse_layers = []
  inp_layers = []
  for i in range(18):
    inp = Input(shape=input_shape, name='horse{}'.format(i))
    inp_layers.append(inp)
    x = Dense(128, activation='relu')(inp)
    x = Dropout(0.2)(x)
    x = Dense(32, activation='relu')(x)
    x = Dropout(0.2)(x)
    horse_layers.append(x)

  concatenated = concatenate(horse_layers)
  x = Dense(128, activation='relu')(concatenated)
  x = Dropout(0.2)(x)
  x = Dense(64, activation='relu')(x)
  x = Dropout(0.2)(x)
  output = Dense(18, activation='softmax')(x)

  model = Model(inp_layers, output)
  model.compile(optimizer='adam', loss='binary_crossentropy')

  return model

## 学習

In [32]:
model = define_race_model((X_train_sc.shape[2],))

In [33]:
model.summary()

Model: "functional_1"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
horse0 (InputLayer)             [(None, 61)]         0                                            
__________________________________________________________________________________________________
horse1 (InputLayer)             [(None, 61)]         0                                            
__________________________________________________________________________________________________
horse2 (InputLayer)             [(None, 61)]         0                                            
__________________________________________________________________________________________________
horse3 (InputLayer)             [(None, 61)]         0                                            
_______________________________________________________________________________________

In [34]:
epoch = 200
batch = 4

In [ ]:
model.fit(X_train_sc_reshape, y_train, epochs=epoch, batch_size=batch)

Epoch 1/200
4908/4908 [==============================] - 29s 6ms/step - loss: 0.4533
Epoch 2/200
4908/4908 [==============================] - 28s 6ms/step - loss: 0.4433
Epoch 3/200
4908/4908 [==============================] - 28s 6ms/step - loss: 0.4407
Epoch 4/200
4908/4908 [==============================] - 28s 6ms/step - loss: 0.4397
Epoch 5/200
4908/4908 [==============================] - 28s 6ms/step - loss: 0.4389
Epoch 6/200
4908/4908 [==============================] - 28s 6ms/step - loss: 0.4376
Epoch 7/200
4908/4908 [==============================] - 28s 6ms/step - loss: 0.4376
Epoch 8/200
4908/4908 [==============================] - 28s 6ms/step - loss: 0.4363
Epoch 9/200
4908/4908 [==============================] - 28s 6ms/step - loss: 0.4364
Epoch 10/200
4908/4908 [==============================] - 28s 6ms/step - loss: 0.4353
Epoch 11/200
4908/4908 [==============================] - 28s 6ms/step - loss: 0.4363
Epoch 12/200
4908/4908 [==============================] - 29s 6

In [ ]:
model.save_weights(DATA_PATH + 'param.hdf5')

In [ ]:
model.save(DATA_PATH + 'multi_input_model')

# 予測

In [ ]:
pred = model.predict(X_test_sc_reshape)

レースごとに予測された確率が最も高い馬が３位以内に入っている確率を算出する。

In [ ]:
def calc_prob(predict):
  """
  レースごとの予測モデルではない場合のスコア計測
  """
  stack_q = 0
  correct = 0
  for query in test_query:
    ind = np.argmax(predict[stack_q:stack_q+query])
    stack_q += query
    if test_df.iloc[ind]['show'] == 1:
      correct += 1

  print('score is', correct / len(test_query))

In [ ]:
def calc_prob_race(true, pred):
  """
  レースごとの予測モデル用スコア計測
  """
  correct = 0
  for i in range(len(true)):
    ind = np.argmax(pred[i])
    if true[i][ind] == 1:
      correct += 1

  print('socre is', correct / len(true))

In [ ]:
calc_prob_race(y_test, pred)

# 考察

## 単純なニューラルネットワークモデル

- 0.6406

## 多入力ニューラルネットワーク

### 出力層をsigmoidにして回帰

- 0.4756

### 出力層をsoftmaxにして分類

- 0.436

#### 多入力側に層を追加

- 0.4246

#### 層を減らして単純に

200epoch

- 0.2836

損失関数が減っていなかった